packages/playwright-test/src/testType.ts

```diff
@@ -80,20 +80,34 @@ export class TestTypeImpl {
       ].join('\n'), location);
       return;
     }
-    if (this._projectSetup !== suite._isProjectSetup) {
-      if (this._projectSetup)
-        addFatalError(`${title} is called in a file which is not a part of project setup.`, location);
-      else
-        addFatalError(`${title} is called in a project setup file (use 'setup' instead of 'test').`, location);
-    }
+    if (allowedContext === 'projectSetup' && !suite._isProjectSetup)
+      addFatalError(`${title} is called in a file which is not a part of project setup.`, location);
+    else if (allowedContext === 'test' && suite._isProjectSetup)
+      addFatalError(`${title} is called in a project setup file (use 'test.projectSetup' instead of 'test').`, location);
     return suite;
   }
 
-  private _createTest(type: 'default' | 'only' | 'skip' | 'fixme', location: Location, title: string, fn: Function) {
+  private _createTest(type: 'default' | 'only' | 'skip' | 'fixme' | 'projectSetup', location: Location, title: string, fn: Function) {
     throwIfRunningInsideJest();
-    const suite = this._currentSuite(location, this._projectSetup ? 'setup()' : 'test()');
+    let functionTitle = 'test()';
+    let allowedContext: 'test' | 'projectSetup' | 'any' = 'any';
+    switch (type) {
+      case 'projectSetup':
+        functionTitle = 'test.projectSetup()';
+        allowedContext = 'projectSetup';
+        break;
+      case 'default':
+        allowedContext = 'test';
+        break
+    }
+    const suite = this._currentSuite(location, functionTitle, allowedContext);
     if (!suite)
       return;
+    if (type === 'projectSetup' && !suite._isProjectSetup)
```

Comment:
Doesn't `_currentSuite` already check this and next conditions?

Reply:
Oops, removed.